File: mergesort/go/mergesort.go
package mergesort

func Sort(a []int) {
	// fmt.Printf("Sort: %v\n", a)

	if len(a) < 2 {
		return
	}

	mid := len(a) / 2

	Sort(a[:mid])
	Sort(a[mid:])
	merge(a, mid)
}

func merge(a []int, mid int) {
	// fmt.Printf("merge: %v with mid: %d\n", a, mid)

	merged := []int{}

	left := 0
	right := mid
	for left < mid || right < len(a) {
		if right >= len(a) || left < mid && a[left] < a[right] {
			merged = append(merged, a[left])
			left++
		} else {
			merged = append(merged, a[right])
			right++
		}
	}

	copy(a, merged)
}

File: mergesort/typescript/mergesort.ts
type CompFn<T> = (a: T, b: T) => number;

const DEFAULT_COMPARE = <T>(a: T, b: T) => a === b ? 0 : a < b ? - 1 : 1;

export default function sort<T>(array: T[], compare = DEFAULT_COMPARE): void {
    sortRec(array, 0, array.length, compare);
}

function sortRec<T>(array: T[], start: number, end: number, compare: CompFn<T>): void {
    const mid = Math.floor(start + (end - start) / 2);

    if (mid === start) {
        return;
    }

    sortRec(array, start, mid, compare);
    sortRec(array, mid, end, compare);
    merge(array, start, mid, end, compare);
}

function merge<T>(array: T[], start: number, mid: number, end: number, compare: CompFn<T>): void {
    const tmp: T[] = [];
    let i = start;
    let j = mid;

    while (tmp.length < end - start) {
        if (i < mid && (j === end || compare(array[i], array[j]) <= 0)) {
            tmp.push(array[i]);
            i += 1;
        } else {
            tmp.push(array[j]);
            j += 1;
        }
    }

    array.splice(start, tmp.length, ...tmp);
}
